infrastructure/merklemountainrange/src/merklemountainrange.rs

```diff
@@ -159,13 +236,36 @@ where
         self.mmr.push(new_node);
         if is_node_right(self.get_last_added_index()) {
             self.add_single_no_leaf(self.get_last_added_index())
+        } else {
+            self.current_peak_height = self.calc_peak_height(); // because we have now stopped adding right nodes, we need to update the height of the mmr
         }
     }
 
     // This function is just a private function to return the index of the last added node
     fn get_last_added_index(&self) -> usize {
         self.mmr.len() - 1
     }
+
+    fn bag_mmr(&self) -> Vec<ObjectHash> {
+        // lets find all peaks of the mmr
+        let mut peaks = Vec::new();
+        self.find_bagging_indexes(self.current_peak_height.0 as i64, self.current_peak_height.1, &mut peaks);
```

Comment:
Feels a little clunky using a mutable reference to act as a return variable, why not just have `find_bagging_indexes(...) -> Vec<ObjectHash>`. That makes it easier to see what the function is actually producing.

Reply:
Look down below, its recursive function. 
Either you need to clone/copy/move the vec each time the function is called, or you just pass in the same vec and keep appending to it.

Reply:
Oh, ok